packages/@lwc/template-compiler/src/parser/expression.ts

```diff
@@ -33,57 +33,62 @@ export function isPotentialExpression(source: string): boolean {
     return !!source.match(POTENTIAL_EXPRESSION_RE);
 }
 
+function validateExpression(
+    node: estree.BaseNode,
+    element: IRNode,
+    allowComputedMemberExpression: boolean
+): node is estree.MemberExpression | estree.Identifier {
+    const isValidNode = node.type === 'Identifier' || node.type === 'MemberExpression';
+    invariant(isValidNode, ParserDiagnostics.INVALID_NODE, [node.type]);
+
+    if (node.type === 'MemberExpression') {
+        const expression = node as estree.MemberExpression;
+
+        invariant(
+            allowComputedMemberExpression || !expression.computed,
+            ParserDiagnostics.COMPUTED_PROPERTY_ACCESS_NOT_ALLOWED
+        );
+
+        // Validate if the expression is modifying an iterator (only the leftmost). Ex: it.next in it.next.foo
+        if (expression.object.type === 'Identifier' && expression.property.type === 'Identifier') {
+            // .object and .property are ensured to be Identifiers.
+            const propertyIdentifier = (expression.property as unknown) as TemplateIdentifier;
+            const objectIdentifier = (expression.object as unknown) as TemplateIdentifier;
+            invariant(
+                !isBoundToIterator(objectIdentifier, element) ||
+                    propertyIdentifier.name !== ITERATOR_NEXT_KEY,
```

Comment:
Let's change this condition to make it more efficient. We should never invoke `isBoundToIterator` if the property is not `next`. 

```ts
invariant(
    propertyIdentifier.name === ITERATOR_NEXT_KEY && isBoundToIterator(objectIdentifier as any, element),
    ParserDiagnostics.MODIFYING_ITERATORS_NOT_ALLOWED
);
```

By doing this, we don't have to optimize `isBoundToIterator` since it will be called WAY less often.